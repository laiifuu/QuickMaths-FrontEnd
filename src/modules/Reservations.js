import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { useNavigate, Link } from 'react-router-dom';

const Reservations = () => {
  const isLoggedIn = JSON.parse(window.localStorage.getItem('logged_in'));
  const navigate = useNavigate();

  useEffect(() => {
    if (!isLoggedIn) {
      setTimeout(() => {
        navigate('/user/login');
      }, 2000);
    }
  }, [isLoggedIn, navigate]);

  const sortReservations = (a, b) => {
    const dateA = new Date(a.date);
    const dateB = new Date(b.date);

    let hoursA = Number(a.hour.replace(/(^\d+)(.+$)/i, '$1'));
    let hoursB = Number(b.hour.replace(/(^\d+)(.+$)/i, '$1'));

    if (a.hour.includes('pm')) {
      hoursA += 12;
    }

    if (b.hour.includes('pm')) {
      hoursB += 12;
    }

    dateA.setHours(hoursA);
    dateB.setHours(hoursB);

    return dateA - dateB;
  };

  let reservations = useSelector((state) => state.users.reservations);
  const tutors = useSelector((state) => state.tutors.tutors);

  if (isLoggedIn) {
    reservations = reservations.sort(sortReservations);
    return (
      <section className="reservations-page">
        <h1>My reservations</h1>
        {reservations.length !== 0 ? (
          <>
            <p className="next-session-info">
              Your next lesson is on
              {' '}
              {((new Date(reservations[0].date)).toUTCString()).substring(0, 16)}
              {' '}
              at
              {' '}
              {reservations[0].hour}
              {' '}
              in
              {' '}
              {reservations[0].city}
              {' '}
              with
              {' '}
              {tutors.find((tutor) => tutor.id === reservations[0].tutor_id).firstName}
            </p>
            <table>
              <thead>
                <tr>
                  <th>Tutor Name</th>
                  <th>City</th>
                  <th>Date</th>
                  <th>Hour</th>
                </tr>
              </thead>
              <tbody>
                {reservations.map((item) => (
                  <tr key={item.id}>
                    <td>
                      {tutors.find((tutor) => tutor.id === item.tutor_id).firstName}
                      {' '}
                      {tutors.find((tutor) => tutor.id === item.tutor_id).lastName}
                    </td>
                    <td>{item.city}</td>
                    <td>{item.date}</td>
                    <td>{item.hour}</td>
                  </tr>
                ))}
              </tbody>
            </table>
          </>
        ) : (
          <div className="no-items-available">
            <div>There are no reservations currently</div>
            <Link to="/reserve"><button type="button">Make a reservation now!</button></Link>
          </div>
        )}
      </section>
    );
  }
  return (
    <div className="popup-message">
      <p>Please log in to access this page</p>
    </div>
  );
};

export default Reservations;
