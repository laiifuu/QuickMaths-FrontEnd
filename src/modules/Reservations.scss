.reservationPageTitle {
  text-align: center;
  padding: 2rem;
}

// .reservationsBody {
//     display: flex;
//     flex-direction: row;
// }

.reservationPageMessage {
  text-align: center;
  padding: 0 3rem;
}

table {
  margin: 3rem auto;
}

td,
th {
  text-align: left;
  padding: 0.5rem;
}

td {
  font-weight: 500;
}

.flexContainer {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100vh;
  padding: 2rem;

  div {
    text-align: center;
  }

  a {
    padding: 1rem;
    background-color: green;
    border-radius: 1rem;
    color: white;
  }
}
