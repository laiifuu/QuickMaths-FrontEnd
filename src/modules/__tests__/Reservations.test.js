import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import store from '../../redux/configureStore';
import Reservations from '../Reservations';

describe('Show all reservations', () => {
  test('Show redirection', async () => {
    const reservations = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Reservations />
        </BrowserRouter>
        ,
      </Provider>,
    )
      .toJSON();
    expect(reservations).toMatchSnapshot();
  });
});
