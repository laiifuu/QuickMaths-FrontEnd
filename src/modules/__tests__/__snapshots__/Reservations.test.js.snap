// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show all reservations Show redirection 1`] = `
Array [
  <div
    className="popup-message"
  >
    <p>
      Please log in to access this page
    </p>
  </div>,
  ",",
]
`;
